### PR TITLE
fix: improve error wrapping with stage-based context in transaction flows

### LINKED ISSUE #122: Improve error wrapping and context propagation in transaction merge and submission flows
## Problem
Current error handling in transaction processing flows (Merge, SubmitTransaction, and SubmitTransactionWithWait) returns basic errors without sufficient context. This makes debugging difficult, especially in distributed environments where tracing failures across components is critical.

For example, errors are often returned directly without indicating:
- Which transaction failed
- Which namespace caused the issue
- At what stage the failure occurred

## Proposed Solution
- Wrap errors using `fmt.Errorf` with `%w` for proper error chaining
- Add contextual information such as:
  - transaction ID (txID)
  - namespace (where applicable)
  - operation stage (merge, submit, wait)
- Ensure consistency in error messages across these flows

## Example
Instead of:
  return err

Use:
  return fmt.Errorf("failed to submit transaction %s: %w", txID, err)

## Scope
- SubmitTransaction (app package)
- SubmitTransactionWithWait (app package)

## Why this matters
Improved error context:
- Makes debugging easier
- Helps in production observability
- Aligns with Go best practices for error handling

## Additional Context
Recent test additions highlighted that while edge cases are handled, error messages lack sufficient detail for tracing failures.

### PR DESCRIPTION
**Pull request by sachin9058**
#### Type of change
- Improvement

#### Description
This PR improves error handling in transaction processing flows by adding clear, stage-based context to error messages.

Changes include:

- **SubmitTransaction**
  - Added contextual error wrapping for:
    - submission preparation
    - transaction broadcast

- **SubmitTransactionWithWait**
  - Added contextual error wrapping for:
    - submission preparation
    - notification client acquisition
    - event subscription
    - transaction broadcast
    - waiting for transaction status

- **Preparation internals**
  - Added contextual error wrapping for:
    - signing identity retrieval
    - orderer client initialization

All errors are wrapped using `%w` to preserve error chaining.  
No functional logic was modified.

#### Additional details (Optional)
- Unit tests for affected packages pass:
  - `go test ./tools/fxconfig/internal/transaction`
  - `go test ./tools/fxconfig/internal/app`

- Integration test failures (missing crypto artifacts) are pre-existing and unrelated to this change.

#### Related issues
Fixes #122

### CODE REVIEW

File: tools/fxconfig/internal/app/submit_test.go [hunk 1]
```diff
@@ -140,6 +146,52 @@ func TestSubmitTransaction_Success(t *testing.T) {
...
+		OrdererProvider: makeOrdererProvider(&mockOrdererClient{}, nil),
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	err := a.SubmitTransaction(ctx, "tx-1", someTx())
+	require.Error(t, err)
+	require.ErrorIs(t, err, context.Canceled)
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/submit_test.go]: should we check that expected error is actually thrown? Otherwise I believe that we only test the mock Broadcast.

File: tools/fxconfig/internal/app/submit_test.go [hunk 2]
```diff
@@ -140,6 +146,52 @@ func TestSubmitTransaction_Success(t *testing.T) {
...
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Nanosecond)
+	defer cancel()
+	time.Sleep(time.Millisecond)
+
+	err := a.SubmitTransaction(ctx, "tx-1", someTx())
+	require.Error(t, err)
+	require.ErrorIs(t, err, context.DeadlineExceeded)
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/submit_test.go]: should we check that expected error is actually thrown? Otherwise I believe that we only test the mock Broadcast.

File: tools/fxconfig/internal/app/submit_test.go [hunk 3]
```diff
@@ -140,6 +146,52 @@ func TestSubmitTransaction_Success(t *testing.T) {
...
+	t.Parallel()
+
+	a := &AdminApp{
+		MspProvider:     makeMSPProvider(&testSigningIdentity{}, nil),
+		OrdererProvider: makeOrdererProvider(&mockOrdererClient{}, nil),
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Nanosecond)
+	defer cancel()
+	time.Sleep(time.Millisecond)
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/submit_test.go]: You could use create a context that is already expired:

```go
ctx, cancel := context.WithTimeout(t.Context(), 0)
```

File: tools/fxconfig/internal/app/submit_test.go [hunk 4]
```diff
@@ -140,6 +146,52 @@ func TestSubmitTransaction_Success(t *testing.T) {
...
 
+func TestSubmitTransaction_ContextCancelled(t *testing.T) {
+	t.Parallel()
+
+	a := &AdminApp{
+		MspProvider:     makeMSPProvider(&testSigningIdentity{}, nil),
+		OrdererProvider: makeOrdererProvider(&mockOrdererClient{}, nil),
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/submit_test.go]:
```suggestion
	ctx, cancel := context.WithCancel(t.Context())
```

File: tools/fxconfig/internal/transaction/merge.go [hunk 5]
```diff
@@ -52,6 +52,7 @@ func sortEndorsementsByMspID(merged *applicationpb.Tx) {
 	}
 }
 
+// TODO: Add test coverage for edge cases in validation.
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/transaction/merge.go]: I believe this should not be part of this PR.

File: tools/fxconfig/internal/transaction/merge.go [hunk 6]
```diff
@@ -27,7 +27,7 @@ func Merge(txs []*applicationpb.Tx) (*applicationpb.Tx, error) {
 	}
 
 	if err := validateTransactionsForMerge(txs); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to merge %d transactions: %w", len(txs), err)
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/transaction/merge.go]: I am wondering if we get extra value from this error message. `validateTransactionsForMerge` already gives detailed information what went wrong during the merging validation.

File: tools/fxconfig/internal/transaction/merge_test.go [hunk 7]
```diff
@@ -110,6 +110,45 @@ func TestMerge_ErrorCases(t *testing.T) {
...
+	tx2.Namespaces[0].ReadWrites = []*applicationpb.ReadWrite{{
+		Key:   []byte("asset-1"),
+		Value: []byte("value-b"),
+	}}
+
+	result, err := Merge([]*applicationpb.Tx{tx1, tx2})
+	require.Error(t, err)
+	require.Nil(t, result)
+	require.Contains(t, err.Error(), "content mismatch")
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/transaction/merge_test.go]: I think having these tests is great - however, there are out of scope for this PR.

File: tools/fxconfig/internal/transaction/merge_test.go [hunk 8]
```diff
@@ -321,3 +360,33 @@ func TestMerge_PreservesTransactionContent(t *testing.T) {
...
+
+	result, err := Merge([]*applicationpb.Tx{tx1, tx2})
+	require.NoError(t, err)
+	require.NotNil(t, result)
+	require.Len(t, result.Endorsements, 1)
+	require.Len(t, result.Endorsements[0].EndorsementsWithIdentity, 3)
+	require.Equal(t, "Org1MSP", result.Endorsements[0].EndorsementsWithIdentity[0].Identity.GetMspId())
+	require.Equal(t, "Org2MSP", result.Endorsements[0].EndorsementsWithIdentity[1].Identity.GetMspId())
+	require.Equal(t, "Org3MSP", result.Endorsements[0].EndorsementsWithIdentity[2].Identity.GetMspId())
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/transaction/merge_test.go]: I think having these tests is great - however, there are out of scope for this PR.

**Review by mbrandenburger — CHANGES_REQUESTED** [review]
Thank you @sachin9058 for creating this PR. A few more comments below.

**Comment by sachin9058** [comment]
@mbrandenburger Thanks for the detailed feedback!

I’ve updated the PR to:
- remove out-of-scope merge tests and TODO comment
- simplify error handling in Merge by removing unnecessary wrapping
- improve test cases with better assertions and updated context handling
- address lint issues

Please let me know if any further changes are needed.

File: tools/fxconfig/internal/app/submit_test.go [hunk 9]
```diff
@@ -25,7 +25,18 @@ type mockOrdererClient struct {
...
+	txID string,
+	_ *applicationpb.Tx,
+) error {
+	if txID == "" {
+		return errors.New("txID is required")
+	}
+	if err := ctx.Err(); err != nil {
+		return err
+	}
 	return m.broadcastErr
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/submit_test.go:29]: I believe we should not modify the mock here to add this extra functionality. Instead, the test should just inject the corresponding error which is then returned as `broadcasstErr`.

File: tools/fxconfig/internal/app/submit_test.go [hunk 10]
```diff
@@ -140,6 +151,53 @@ func TestSubmitTransaction_Success(t *testing.T) {
...
+	}
+
+	ctx, cancel := context.WithTimeout(t.Context(), 0)
+	defer cancel()
+
+	err := a.SubmitTransaction(ctx, "tx-1", someTx())
+	require.Error(t, err)
+	require.ErrorIs(t, err, context.DeadlineExceeded)
+	require.Contains(t, err.Error(), "broadcast")
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/submit_test.go]: Actually, we can remove this test as we cover it already with `TestSubmitTransaction_ContextCancelled`

File: tools/fxconfig/internal/app/submit_test.go [hunk 11]
```diff
@@ -140,6 +151,53 @@ func TestSubmitTransaction_Success(t *testing.T) {
...
+
+	a := &AdminApp{
+		MspProvider:     makeMSPProvider(&testSigningIdentity{}, nil),
+		OrdererProvider: makeOrdererProvider(&mockOrdererClient{}, nil),
+	}
+
+	err := a.SubmitTransaction(t.Context(), "", someTx())
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "txID is required")
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/submit_test.go]: This test only tests the mock not the actual app logic :(

**Review by mbrandenburger — CHANGES_REQUESTED** [review]
Thank you @sachin9058 for improving this PR. See my comments below.

**Comment by sachin9058** [comment]
@mbrandenburger Thanks for the feedback!

I’ve updated the tests to:
- simplify the mock behavior
- remove redundant tests
- ensure tests validate actual application logic rather than mock behavior

Please let me know if anything else should be adjusted.

File: tools/fxconfig/internal/app/submit_test.go (hunk 9 above)
Comment (mbrandenburger) [reply tools/fxconfig/internal/app/submit_test.go:29]: reformatting not needed anymore.

**Review by mbrandenburger — CHANGES_REQUESTED** [review]

**Comment by sachin9058** [comment]
Thanks! I’ve reverted the formatting change to keep the diff minimal.

**Review by mbrandenburger — APPROVED** [review]
Thank you @sachin9058
